### PR TITLE
EC2 Instance Connect Endpoint: add aws metadata to Nodes

### PR DESCRIPTION
**Pull request by marcoandredinis**
This PR adds AWS metadata to the Server resource.
This metadata is required for servers of subkind OpenSSHEphemeralKey.

Issue: https://github.com/gravitational/teleport/pull/29316

### CODE REVIEW

File: api/types/constants.go [hunk 1]
```diff
@@ -138,6 +138,16 @@ const (
...
 	SubKindOpenSSHNode = "openssh"
 
+	// SubKindOpenSSHEphemeralKeyNode is a registered OpenSSH (agentless) node that doesn't require trust in Teleport CA.
+	// For each session an SSH Key is created and uploaded to the target host using a side-channel.
+	//
+	// For Amazon EC2 Instances, it uploads the key using:
+	// https://docs.aws.amazon.com/ec2-instance-connect/latest/APIReference/API_SendSSHPublicKey.html
+	// This Key is valid for 60 seconds.
+	//
+	// It uses the private key to SSH into the host.
```
Comment (codingllama) [inline api/types/constants.go:148]: Whose private key?

File: api/types/server.go [hunk 2]
```diff
@@ -119,6 +123,24 @@ func NewServerWithLabels(name, kind string, spec ServerSpecV2, labels map[string
 	return server, nil
 }
 
+// NewNode is a convenience method to create a Server of Kind Node.
+// It accepts a SubKind, ServerSpec and labels.
```
Comment (codingllama) [inline api/types/server.go:127]: Maybe:

```suggestion
```

It's all the signature.

File: api/types/server.go [hunk 3]
```diff
@@ -397,6 +419,64 @@ func (s *ServerV2) setStaticFields() {
...
+// This returns true for SubKindOpenSSHNode and SubKindOpenSSHEphemeralKeyNode.
+func (s *ServerV2) IsOpenSSHNode() bool {
+	return s.SubKind == SubKindOpenSSHNode || s.SubKind == SubKindOpenSSHEphemeralKeyNode
+}
+
+// openSSHNodeCheckAndSetDefaults are common validations for OpenSSH nodes.
+// They include SubKindOpenSSHNode and SubKindOpenSSHEphemeralKeyNode.
+func (s *ServerV2) openSSHNodeCheckAndSetDefaults() error {
+	if s.Spec.Addr == "" {
+		return trace.BadParameter(`Addr must be set when server SubKind is "openssh"`)
```
Comment (codingllama) [inline api/types/server.go:432]: nit: we usually avoid starting Go errors with uppercase letters.

Same for others.

https://github.com/golang/go/wiki/CodeReviewComments#error-strings

File: api/types/server.go [hunk 4]
```diff
@@ -397,6 +419,64 @@ func (s *ServerV2) setStaticFields() {
...
+}
+
+// openSSHEphemeralKeyNodeCheckAndSetDefaults are validations for SubKindOpenSSHEphemeralKeyNode.
+func (s *ServerV2) openSSHEphemeralKeyNodeCheckAndSetDefaults() error {
+	if err := s.openSSHNodeCheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Currently, only AWS EC2 is supported for EphemeralKey mode.
+	if s.Spec.CloudMetadata == nil || s.Spec.CloudMetadata.AWS == nil {
```
Comment (codingllama) [inline api/types/server.go:455]: Suggestion: this is a good fit for a switch:

```go
switch {
case s.Spec.CloudMetadata == nil || s.Spec.CloudMetadata.AWS == nil:
  //
case s.Spec.CloudMetadata.AWS.AccountID == "":
  //
// so on ...

}

File: api/types/server_test.go [hunk 5]
```diff
@@ -354,6 +357,268 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
+					Addr:     "example:22",
+					Hostname: "openssh-node",
+				},
+			},
+			assertion: func(t *testing.T, s *ServerV2, err error) {
+				require.EqualError(t, err, `AWS CloudMetadata is required for "openssh-ephemeral-key" SubKind`)
+			},
+		},
+		{
+			name: "OpenSSHEphemeralKey node without cloud metadata but missing aws info",
```
Comment (codingllama) [inline api/types/server_test.go:380]:
```suggestion
			name: "OpenSSHEphemeralKey node with cloud metadata but missing aws info",
```

I think?

File: api/types/server_test.go [hunk 6]
```diff
@@ -354,6 +357,268 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
+						},
+					},
+				},
+			},
+			assertion: func(t *testing.T, s *ServerV2, err error) {
+				require.EqualError(t, err, `AWS OIDC Integration is required for "openssh-ephemeral-key" SubKind`)
+			},
+		},
+		{
+			name: "OpenSSHEphemeralKey node with aws cloud metadata but missing accountid",
```
Comment (codingllama) [inline api/types/server_test.go:568]: Wrong description?

File: api/types/server_test.go [hunk 7]
```diff
@@ -354,6 +357,268 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
+								AvailabilityZone: "us-east-1b",
+								VPCID:            "vpc-abcd",
+								Integration:      "teleportdev",
+							},
+						},
+					},
+				}
+				require.Equal(t, expectedServer, s)
+
+				require.True(t, s.IsOpenSSHNode(), "IsOpenSSHNode must be true for this node")
```
Comment (codingllama) [inline api/types/server_test.go:619]:
```suggestion
				assert.Equal(t, expectedServer, s)
				assert.True(t, s.IsOpenSSHNode(), "IsOpenSSHNode must be true for this node")
```

https://github.com/golang/go/wiki/TestComments#keep-going

File: api/types/server_test.go [hunk 8]
```diff
@@ -354,6 +357,268 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
+						AWS: &AWSInfo{
+							InstanceID:       "i-123456789012",
+							Region:           "us-east-1",
+							AvailabilityZone: "us-east-1b",
+							VPCID:            "vpc-abcd",
+							Integration:      "teleportdev",
+						},
+					},
+				},
+			},
```
Comment (codingllama) [inline api/types/server_test.go:422]: Suggestion: for tests like these, I like starting with a valid instance and changing it to cause the wanted error.

For example, consider something like this:

```suggestion
			server: makeOpenSSHEmphemeral(func (s *ServerV2) {
				s.Spec.CloudMetadata.AccountID = ""
			}),
```

I'll often declare the equivalent to `makeOpenSSHEmphemeral` in the body of the test itself, as in:

```go
makeOpenSSHEmphemeral = func(fn func(s *ServerV2)) *ServerV2 {
	s := ... // Declare a valid server instance here.
	fn(s)
	return s
}
```

Many ways to do it.

File: lib/web/ui/server.go [hunk 9]
```diff
@@ -82,15 +84,20 @@ func MakeServers(clusterName string, servers []types.Server, accessChecker servi
...
 			Name:        server.GetName(),
 			Hostname:    server.GetHostname(),
 			Addr:        server.GetAddr(),
 			Tunnel:      server.GetUseTunnel(),
 			SSHLogins:   serverLogins,
-		})
+		}
+		if server.GetCloudMetadata() != nil {
+			s.AWS = server.GetCloudMetadata().AWS
+		}
```
Comment (codingllama) [inline lib/web/ui/server.go:98]: Are we guaranteed a single server with AWS data? Is it intended for the last one to win?

Could we have a comment here explaining it, please?

File: api/types/constants.go (hunk 1 above)
Comment (marcoandredinis) [reply api/types/constants.go:148]: We create an SSH Key and upload it to the EC2 instance.
We then use the private key (that was created above) to access the EC2 instance over ssh.

I guess this was not very clear :thinking: 
Should I rephrase?

File: api/types/server.go (hunk 3 above)
Comment (marcoandredinis) [reply api/types/server.go:432]: Yeah, I saw that as well
This code already existed but I'm going to downcase the starting word :+1:

File: api/types/server_test.go (hunk 5 above)
Comment (marcoandredinis) [reply api/types/server_test.go:380]: Yeah :pray:

File: api/types/server_test.go (hunk 6 above)
Comment (marcoandredinis) [reply api/types/server_test.go:568]: Yeah :+1:

File: api/types/server_test.go (hunk 7 above)
Comment (marcoandredinis) [reply api/types/server_test.go:619]: Switched to assert

File: lib/web/ui/server.go (hunk 9 above)
Comment (marcoandredinis) [reply lib/web/ui/server.go:98]: Sorry, I'm not sure I understood the question
Are you asking if each server instance will match a single AWS EC2 instance?
If that's the question, then yes :+1:

File: api/types/constants.go (hunk 1 above)
Comment (codingllama) [reply api/types/constants.go:148]: I think if it said "It uses the private key created above to SSH into the host" I'd already get it.

File: api/types/server.go (hunk 3 above)
Comment (codingllama) [reply api/types/server.go]: If you could rephrase it so "AWS" is not the first word it would be better.

File: lib/web/ui/server.go (hunk 9 above)
Comment (codingllama) [reply lib/web/ui/server.go:98]: It makes no sense because I misread the loop. Looks good, don't mind me. :)

**Review by codingllama — COMMENTED** [review]
Looks good, just waiting for you to address and/or close the remaining comments.

File: api/types/server.go [hunk 10]
```diff
@@ -451,26 +451,27 @@ func (s *ServerV2) openSSHEphemeralKeyNodeCheckAndSetDefaults() error {
...
-		return trace.BadParameter("AWS InstanceID is required for %q SubKind", s.SubKind)
-	}
-	if s.Spec.CloudMetadata.AWS.AvailabilityZone == "" {
-		return trace.BadParameter("AWS Availability Zone is required for %q SubKind", s.SubKind)
-	}
-	if s.Spec.CloudMetadata.AWS.VPCID == "" {
-		return trace.BadParameter("AWS VPC ID is required for %q SubKind", s.SubKind)
+	switch {
+	case s.Spec.CloudMetadata == nil || s.Spec.CloudMetadata.AWS == nil:
+		return trace.BadParameter("missing AWS CloudMetadata (required for %q SubKind)", s.SubKind)
```
Comment (codingllama) [inline api/types/server.go:456]: Nicely done :)

File: api/types/server_test.go [hunk 11]
```diff
@@ -279,7 +307,7 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
 				},
 			},
 			assertion: func(t *testing.T, s *ServerV2, err error) {
-				require.EqualError(t, err, `Addr must be set when server SubKind is "openssh"`)
+				require.EqualError(t, err, `addr must be set when server SubKind is "openssh"`)
```
Comment (codingllama) [inline api/types/server_test.go:310]: Might I suggest `require.ErrorContains`, so the test is less brittle?

```suggestion
				require.ErrorContains(t, err, "addr must be set")
```

(Or maybe even just "addr" would be enough.)

Same for others.

File: api/types/server_test.go [hunk 12]
```diff
@@ -562,34 +493,12 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
-							Region:           "us-east-1",
-							AvailabilityZone: "us-east-1b",
-							VPCID:            "vpc-abcd",
-							Integration:      "teleportdev",
-						},
-					},
-				},
-			},
+			name:   "valid OpenSSHEphemeralKey node",
+			server: makeOpenSSHEphemeralKeyNode(func(_ *ServerV2) {}),
```
Comment (codingllama) [inline api/types/server_test.go:501]:
```suggestion
			server: makeOpenSSHEphemeralKeyNode(func(*ServerV2) {}),
```

We could also make it work with `nil`. Up to you.

File: api/types/server_test.go [hunk 13]
```diff
@@ -360,153 +388,56 @@ func TestServerCheckAndSetDefaults(t *testing.T) {
...
-					Name:      "5da56852-2adb-4540-a37c-80790203f6a9",
-					Namespace: defaults.Namespace,
-				},
-				Spec: ServerSpecV2{
-					Addr:     "example:22",
-					Hostname: "openssh-node",
-				},
-			},
+			server: makeOpenSSHEphemeralKeyNode(func(s *ServerV2) {
+				s.Spec.CloudMetadata = nil
```
Comment (codingllama) [inline api/types/server_test.go:392]: This is much simpler to reason about, thanks.

**Review by codingllama — APPROVED** [review]
LGTM.

**Review by codingllama — APPROVED** [review]

**Review by codingllama — APPROVED** [review]

File: api/types/constants.go [hunk 14]
```diff
@@ -138,6 +138,16 @@ const (
...
 
+	// SubKindOpenSSHEphemeralKeyNode is a registered OpenSSH (agentless) node that doesn't require trust in Teleport CA.
+	// For each session an SSH Key is created and uploaded to the target host using a side-channel.
+	//
+	// For Amazon EC2 Instances, it uploads the key using:
+	// https://docs.aws.amazon.com/ec2-instance-connect/latest/APIReference/API_SendSSHPublicKey.html
+	// This Key is valid for 60 seconds.
+	//
+	// It uses the private key created above to SSH into the host.
+	SubKindOpenSSHEphemeralKeyNode = "openssh-ephemeral-key"
```
Comment (greedy52) [inline api/types/constants.go:149]: wonder if we should just make a subkind for ec2 instance connect explicitly. If we have another type of `SubKindOpenSSHEphemeralKeyNode` in the future, what fields/attributes we will use to differentiate between those?

**Review by greedy52 — APPROVED** [review]
LGTM

File: api/types/constants.go (hunk 14 above)
Comment (marcoandredinis) [reply api/types/constants.go:149]: I can rename it to OpenSSHEC2InstanceConnectEndpoint (`openssh-ec2-ice`)
What do you think?

Comment (greedy52) [reply api/types/constants.go:149]: > I can rename it to OpenSSHEC2InstanceConnectEndpoint (openssh-ec2-ice)

👍 

@codingllama what's your thought on an explicit subkind like this vs a slightly more generic `SubKindOpenSSHEphemeralKeyNode`?

Comment (codingllama) [reply api/types/constants.go:149]: That's alright by me, you are the experts as far I'm concerned. (It does seem like a good idea.)

**Pull request merged**